akka-stream/src/main/scala/akka/stream/impl/fusing/AggregateWithTimeBoundary.scala

```diff
@@ -0,0 +1,133 @@
+ */
+
+package akka.stream.impl.fusing
+
+import akka.stream.stage._
+import akka.stream.{ Attributes, FlowShape, Inlet, Outlet }
+
+import scala.concurrent.duration._
+
+/**
+ * This is a more scalable and general case of [[GroupedWeightedWithin]]
+ * which groups a stream into vectors based on custom weight and time.
+ * The problem with that solution is each grouped vector must fit into memory before emitting to the next stage.
+ * That won't work for the use case of writing large files.
+ * The desirable behavior is to write data as they come as opposed to accumulate everything until the group slicing condition is met.
+ * In this case, the output channel needs to be closed if there is no data arriving within certain time to avoid connection timeout.
+ * This custom flow uses custom aggregator to support such use cases.
+ * Upstream inputs are continuously aggregated as they arrive.
+ * The currently aggregated elements can be terminated and emitted based on custom conditions
+ *
+ * @param allocate    allocate the initial data structure for aggregated elements
+ * @param aggregate   update the aggregated elements, return true if ready to emit after update
+ * @param emitOnTimer decide whether the currently aggregated elements can be emitted on each timer event
+ * @param harvest     this is invoked before emit within the current stage/operator
+ */
+class AggregateWithBoundary[T, Agg, Emit](
+    allocate: => Agg,
+    aggregate: (Agg, T) => Boolean,
+    harvest: Agg => Emit,
+    emitOnTimer: Option[(Agg => Boolean, FiniteDuration)])
+    extends GraphStage[FlowShape[T, Emit]] {
+
+  emitOnTimer.foreach {
+    case (_, interval) => require(interval.gteq(1.milli), s"timer(${interval.toCoarsest}) must not be smaller than 1ms")
+  }
+
+  val in: Inlet[T] = Inlet[T](s"${this.getClass.getName}.in")
+  val out: Outlet[Emit] = Outlet[Emit](s"${this.getClass.getName}.out")
+  override val shape: FlowShape[T, Emit] = FlowShape(in, out)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new TimerGraphStageLogic(shape) with InHandler with OutHandler {
+
+      private[this] var aggregated: Agg = null.asInstanceOf[Agg]
+
+      override def preStart(): Unit = {
+        emitOnTimer.foreach {
+          case (_, interval) => scheduleWithFixedDelay(s"${this.getClass.getSimpleName}Timer", interval, interval)
+        }
+      }
+
+      override protected def onTimer(timerKey: Any): Unit = {
+        emitOnTimer.foreach {
+          case (isReadyOnTimer, _) => if (aggregated != null && isReadyOnTimer(aggregated)) harvestAndEmit()
+        }
+      }
+
+      // at onPush, isAvailable(in)=true hasBeenPulled(in)=false, isAvailable(out) could be true or false
+      override def onPush(): Unit = {
+        val input = grab(in)
+        if (aggregated == null) aggregated = allocate
+        if (aggregate(aggregated, input)) harvestAndEmit()
+        if (isAvailable(out)) pull(in)
```

Comment:
Actually, this could pull immediately if aggregate returned false and wants to continue aggregating, even if out is not available, until the aggregate is `complete`, if at that time downstream is still backpressuring we need to propagate backpressure, until the complete aggregate has been emitted.

There is a callback that can be used with emit to trigger some logic once it has emitted that can be used for that.

That would also work nicely with an early pull in `preStart`

Reply:
If `aggregate` return false, `out` should always be available at this point except if the timer has triggered an emit.
We do not want to pull in that case either, since the timer can keep emitting things, while onPush keep pulling here, this can "leak" the back pressure indefinitely in the worst case.